public/app/plugins/datasource/cloudwatch/datasource.ts

```diff
@@ -613,7 +613,6 @@ export class CloudWatchDatasource extends DataSourceApi<CloudWatchQuery, CloudWa
           return {
             error: error || queryResult.error ? { message: queryResult.error } : null,
             data: [
-              ...data,
```

Comment:
this whole reduce function was very strange and hard to understand, seems like it can be simplified and some comments added to what is being done. Looping through the results and then looking up the queries might be an easier more intuitive way to write this. 

The reduce method on the queries and then for each query rebuild the whole data array was really messy and hard to follow for me, especially with no comments that explain what the code is trying to do

Reply:
exactly my thoughts. Let me change this then.

Reply:
see frontend chan I think this function looks buggy, there seems to be no matching between query and result (what data frame the link is added to, it adds it to all, so the last query will overwrite all), if I read the code correctly

Reply:
@zoltanbedi 
here is the solution I had prepared (I wasn't so much sure about the error handling, therefore I tried to imitate the existing approach):
```
  async performTimeSeriesQuery(request: MetricRequest, { from, to }: TimeRange): Promise<any> {
    try {
      const res: TSDBResponse = await this.awsRequest(TSDB_QUERY_ENDPOINT, request);
      const dataframes: DataFrame[] = toDataQueryResponse({ data: res }).data;
      if (!dataframes || dataframes.length <= 0) {
        return { data: [] };
      }

      const data = dataframes.map(frame => {
        const queryResult = res.results[frame.refId!];
        const error = queryResult.error ? { message: queryResult.error } : null;
        if (!queryResult) {
          return { frame, error };
        }

        const requestQuery = request.queries.find(q => q.refId === frame.refId!) as any;

        const link = this.buildCloudwatchConsoleUrl(
          requestQuery!,
          from.toISOString(),
          to.toISOString(),
          frame.refId!,
          queryResult.meta.gmdMeta
        );

        if (link) {
          for (const field of frame.fields) {
            field.config.links = [
              {
                url: link,
                title: 'View in CloudWatch console',
                targetBlank: true,
              },
            ];
          }
        }
        return { frame, error };
      });

      return {
        data: data.map(o => o.frame),
        error: data
          .map(o => o.error)
          .reduce((err, error) => {
            return err || error;
          }, null),
      };
    } catch (err) {
      if (/^Throttling:.*/.test(err.data.message)) {
        const failedRedIds = Object.keys(err.data.results);
        const regionsAffected = Object.values(request.queries).reduce(
          (res: string[], { refId, region }) =>
            (refId && !failedRedIds.includes(refId)) || res.includes(region) ? res : [...res, region],
          []
        ) as string[];

        regionsAffected.forEach(region => this.debouncedAlert(this.datasourceName, this.getActualRegion(region)));
      }

      if (err.data && err.data.message === 'Metric request error' && err.data.error) {
        err.data.message = err.data.error;
      }

      throw err;
    }
  }
```

Reply:
that looks way more reasonable & readable and looks you here match the result with the correct query and not just the last query

Reply:
Thanks Sofia, this looks good.